planner/core/logical_plan_builder.go

```diff
@@ -1372,10 +1373,11 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
 			childTp := u.children[j].Schema().Columns[i].RetType
 			resultTp = unionJoinFieldType(resultTp, childTp)
 		}
-		if err := expression.CheckIllegalMixCollation("UNION", tmpExprs, types.ETInt); err != nil {
-			return err
+		dstCharset, dstCollation, coercibility, err := expression.CheckAndDeriveCollationFromExprsWithCoer(b.ctx, "UNION", resultTp.EvalType(), tmpExprs...)
```

Comment:
@yuqi1129 here is the fix, aggregate `date(binary)` and `int(binary)` to `string` will have charset and collation according to `character_set_collation` and `collation_connection`.